src/Types init.

otruffer opened the pull request:
Branch is on the ILIAS repo, sorry for that! I can't move it anymore, because the discussion has already started. I will delete the branch after a merge/rejection.

otruffer committed:
src/Types init.

Amstutz commented on src/Types/Exceptions/Exception.php:
```diff
@@ -0,0 +1,16 @@
+ *
+ * @package ILIAS\Types
+ *
+ *          The Basic Exception Class for BackgroundTasks. Please Specify by extending
```
Indentation seems off. Add period at the end.

Amstutz commented on src/Types/Exceptions/InvalidArgumentException.php:
```diff
@@ -0,0 +1,13 @@
+ *
+ * @package ILIAS\Types
+ *
+ *          A bucket that contains a user interaction needs at least one observer.
```
Indentation seems off.

Amstutz commented on src/Types/ListType.php:
```diff
@@ -0,0 +1,128 @@
+
+use ILIAS\Types\Exceptions\InvalidArgumentException;
+
+class ListType implements Type {
```
Add a sentence here describing, what this type is about or implement a factory generating  instances of this type and add the description there.

Amstutz commented on src/Types/ListType.php:
```diff
@@ -0,0 +1,128 @@
+	 */
+	public function __construct($fullyQualifiedClassName) {
+		if(!is_a($fullyQualifiedClassName, Type::class))
+			$fullyQualifiedClassName = new SingleType($fullyQualifiedClassName);
```
I prefer curly brackets. Maybe we should add this explicitly to:
 [http://www.ilias.de/docu/goto_docu_pg_202_42.html](http://www.ilias.de/docu/goto_docu_pg_202_42.html). What do the others think?

otruffer replied:
I prefer as little brackets as reasonable => less clutter.

klees replied:
I like reducing clutter, but would support @Amstutz on this one. Having brackets after every `if` helps reducing bugs when adding additional statements to the block of the `if`. At least common dev folklore says so...

Amstutz commented on src/Types/ListType.php:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	/**
+	 * @return string A string representation of the Type.
```
Use @inheritdoc here.

Amstutz commented on src/Types/ListType.php:
```diff
@@ -0,0 +1,128 @@
+	 * @param $type Type
+	 * @return bool
+	 */
+	function isSubtypeOf(Type $type) {
```
Use @inheritdoc here.

Amstutz commented on src/Types/ListType.php:
```diff
@@ -0,0 +1,128 @@
+	 */
+	function isSubtypeOf(Type $type) {
+		if(!$type instanceof ListType)
+			return false;
```
Curly brackets again.

Amstutz commented on src/Types/ListType.php:
```diff
@@ -0,0 +1,128 @@
+	 * @return Type
+	 * @throws InvalidArgumentException
+	 */
+	public static function calculateLowestCommonType($list) {
```
I do not like static methods due to bad testability (see e.g. [http://misko.hevery.com/2008/12/15/static-methods-are-death-to-testability/](http://misko.hevery.com/2008/12/15/static-methods-are-death-to-testability/)). Reconsider this.

klees replied:
I think this is not used anywhere. Or is my Ctrl-F broken?

Amstutz commented on src/Types/ListType.php:
```diff
@@ -0,0 +1,128 @@
+	 * @throws InvalidArgumentException
+	 */
+	public static function calculateLowestCommonType($list) {
+		// If the list is empty the type should be [] (empty list).
```
But you return null. I think null is better, since type is not a list. Also check if the return type in the doc comments is correct.

Amstutz commented on src/Types/VoidType.php:
```diff
@@ -0,0 +1,61 @@
+namespace ILIAS\Types;
+
+
+class VoidType implements Type {
```
Add a sentence here describing, what this type is about or implement a factory generating instances of this type and add the description there.

klees replied:
Do I understand correctly that this is different from `null` in the regard that it does not contain any value?

Amstutz commented on src/Types/VoidType.php:
```diff
@@ -0,0 +1,61 @@
+	/**
+	 * @return VoidType
+	 */
+	public static function instance(){
```
Why not moving this out to some factory? I do not like this pattern too much.

otruffer replied:
A void type seems to be one of the few valid cases for a singleton. What's the advantage of a factory? Multiple instances of a void type? A global void factory? Sounds like overengineering to me.

Amstutz commented on src/Types/TupleType.php:
```diff
@@ -0,0 +1,72 @@
+	 * @return bool
+	 */
+	function equals(Type $otherType) {
+		// TODO: Implement equals() method.
```
Why is this not implemented?

Amstutz commented on src/Types/TupleType.php:
```diff
@@ -0,0 +1,72 @@
+	 * @return Type[]
+	 */
+	function getAncestors() {
+		// TODO: Implement getAncestors() method.
```
Why is this not implemented?

Amstutz commented on src/Types/Type.php:
```diff
@@ -0,0 +1,34 @@
+	function __toString();
+
+	/**
+	 * Is this type a subtype of $type. Not strict! x->isSubtype(x) == true.
```
I am afraid, this needs some more explanation. When exactly is a type a subtype of another? Especially for tuples and lists, this is not immediately clear to me.

klees replied:
I second the comment of @Amstutz. What can I expect if one type is a subtype of another type? If this is intended to be understood in the "subclass" sense, the implementation of this method for Lists is incorrect, as explained [here](http://stackoverflow.com/a/2745301).

Amstutz commented on src/Types/README.md:
```diff
@@ -0,0 +1,37 @@
+
+UnitTests
+---------
+Unit tests are currently coupled with BackgroundTasks, they will be added later.
```
Please add them now.

Amstutz commented on src/Types/ListType.php:
```diff
@@ -0,0 +1,128 @@
+		}
+
+		$lct = $ancestorsList[0][0];
+		foreach ($ancestorsList[0] as $i => $ancestors) {
```
Should this not be, $i => $ancestor (not ancenstor**s**)?

Amstutz commented on src/Types/ListType.php:
```diff
@@ -0,0 +1,128 @@
+	 */
+	protected static function sameClassOnLevel($ancestorsList, $i) {
+		$class = $ancestorsList[0][$i];
+		foreach($ancestorsList as $class_hierarchy) {
```
Is there a way for excluding $ancestorsList[0] from this loop? Seems useless, to test against the origin of $class itself.

Amstutz commented on src/Types/ListType.php:
```diff
@@ -0,0 +1,128 @@
+
+	/**
+	 * Todo: This implementation is not performing well (needs the most iterations) on lists with all the same type,
+	 * this might be suboptimal.
```
Currently we have O(n^2), this is only critical if we have a huge list or many ancestors. I think this is rarely the case. What you could do, if you think, performance is crucial is reverse the checking order (start from the topmost ancestor) and then store and omit tests for equal lists. E.g. if you have three listTypes (A,B and C), two (A, B) of them with identical ancestors you could discard the checks for B after the first round, since you know, that they are all identical to A. However, not sure if this increased complexity is worth the extra effort.

Amstutz reviewed (changes requested):
Hi @otruffer ,

Thank you a lot for this PR. Thanks for the readme.

I think this goes in a good direction, however I have things to nag about:

- Use @inheritdoc to avoid redundancies in comments
- Add curly brackets after if's (maybe only personal preference), if others disagree with this, please scream.
- Add Unit Tests
- Do not use static methods if there is any other option (which probably means, do not use static methods).

Finally there is one last point, which might be a matter personal preference. I am curious what other developers think of this. I heavily recommend to make a clear distinction between specification and implementation. See (from [https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/src/README.md](https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/src/README.md)):
`7. All libraries MUST expose a clear public interface to their consumers. Every class, interface or trait on the top level of the library is considered to be a part of the public interface, as well as all entities that are reachable via entities on the top level.`

Maybe throw all classes into an implementation folder and clearly label their public interfaces through php interfaces. I would also recommend to use a factory and put the documentation of the different type into the interface of this factory.

klees commented:
Hi @otruffer, I assigned you since you are in charge of this PR, I guess. Will have a look tomorrow.

klees commented on src/Types/Type.php:
```diff
@@ -0,0 +1,34 @@
+	 * @param Type $otherType
+	 * @return bool
+	 */
+	function equals(Type $otherType);
```
How is that different from `==` or `===`?

otruffer replied:
it's OO

klees reviewed (changes requested):
Hi @otruffer, I have this subtype-issue...

otruffer committed:
refactor of types.

otruffer commented:
Thanks for your comments. Major changes:
- calculateLowestCommonType is removed. It was an artifact of other development activities.
- isSubtypeOf renamed to isExtensionOf and added doc that the behaviour is covariant. I emulate the java generic behaviour of <? extends A>. 
- Added tests for types.
- Extraction of getAncestors in a different interface. Tuples do not behave nicely with this method.

Major non-changes:
- Void stays a singleton
- BT wants to state that it does not return a VoidType (explicitly does NOT return anything) . Which will behave way more consistent than to return null. You can invoke the expected methods on VoidType without getting a nullpointer etc.
- The implementations of Single- List- and TupleTypes stay on top level. A public interface of only Type seems uncomfortable for me. We might extract Single List and Tuple as interfaces with no additional methods?

Amstutz commented:
Still not a big fan of the way the types will be instantiated (would prefer a factory). But I think it's fine for me, this might be rather personal taste.

klees commented on src/Types/ListType.php:
```diff
@@ -0,0 +1,104 @@
+	 * Attention: [Dog].isExtensionOf([Animal]) == true. In other words:
+	 * The isExtensionOf behaves Covariant!
+	 *
+	 * If you are familiar with e.g. Java Generics ?.isExtensionOf(x) behaves the same as <? extends x>.
```
Hi Oskar,

I don't see how this comment solves this problem in any way. `SingleType::isExtensionOf` implements this method by using `isSubclassOf`, which conveys a different meaning than the one announced here. This implementation is also different from the interface defined in `Type`. There is no meaningful way to implement a `List of (subclass of X)s` as a subclass of `List of Xs`. The link in the comment explains that properly. I thus don't see which information I (as a user of Type) get, if I call `isExtensionOf` on a type object.

Best regards!

otruffer replied:
I partially agree:
>  There is no meaningful way to implement a List of (subclass of X)s as a subclass of List of Xs

Is certainly true, thus I did not call it isSubclassOf but isExtensionOf. The difference is what I tried to explain in the commentary: isSubclassOf (not implemented here) would behave covariant on SingleTypes and contravariant on ListTypes. isExtensionOf behaves covariant in both the SingleType and the List Type. With the one crucial effect:

new ListType(Dog::class).isExtensionOf(new ListType(Animal::class)) == true

Thus it is also valid to implement isExtensionOf on SingleTypes using the reflection class' isSubclass, as on the SingleType they both behave covariant.

Why I think the covariant implementation is more useful atm (we can have both later) than the contravariant, and why I use the covariant in the background tasks (this is open for discussion): 
If I have a job that accepts ILIAS Objects "ListType(ILIASObject)" and gives back their ref ids "ListType(IntegerValue)" I want the Job to also be able to handle "ListType(ilObjCourse)", as getting the ref ids of different ilObjects would be tedious, thus we need the covariant implementation somehow. 

To be perfectly in sync with e.g. Java we would need an additional typelike helper Extends(_). Then we could our ref id job to take an argument of type ListType(Extends(SingleType(ilObject))). This would be similar to:

[1] List<Animal>  <--> ListType(SingleType(Animal))
[2] List<? extends Animal>  <-->  ListType(Extends(SingleType(Animal)))

Where the isSubclassOf behaves contravariant for [1] and covariant for [2]. We might want to implement isSubclassOf later on, or do you think this is urgent? Thus follows for your question:

> I thus don't see which information I (as a user of Type) get, if I call isExtensionOf on a type object.

You always get covariant behaviour. And yes, this is a concern for type safety.

Please let me know if that makes sense to you and whether you agree with this approach.

Cheers!

Edit: Formatting.

klees replied:
Hi Oskar,
I really see your use case but I still disagree with the implementation. Let's try again =)

I think the problem boils down to these two problems:
* If I have any two types `A` and `B` the only things I could do with them are defined via the interface `Type`. If `A.isExtensionOf(B) === true`: What does it tell me? How does it help? I certainly cannot treat the `A` like a `B`, because this exactly breaks down in the list case. Thus, if I have a background task (or other thingy) acting on a `B`, I cannot feed an `A` into it.
* What is your understanding of covariant and contravariant? As far as I understand, List is neither covariant nor contravariant, hence the problem. `pop` is covariant. If I have a list that contains `A`s and pop values from it, I could treat the `A`s as `B`s if `A.isExtensionOf(B)`. `push` on the other hand is contravariant. If I have a List of `B`s I could push `A`s into it if `A.isExtensionOf(B)`, which is exactly the other way round. If I had a list of `A`s I cannot push `B`s into it, even if `A.isExtensionOf(B)`, because an `A` is somehow more than a `B`.

My goal is not to be in sync with Java, because Java does it wrong, as far as I understand, and is unsound there.

The idea with the `Extends`-type seems to be sound, though.

And also nice to have a discussion on this level here.

Best Regards!

otruffer replied:
Hi Richard

Notice: In the following if you read A<'T> you can ignore the tick just before the T. It's only there to prevent Github from sanitizing the whole <....> part.

> If I have any two types A and B the only things I could do with them are defined via the interface Type. If A.isExtensionOf(B) === true: What does it tell me? How does it help? I certainly cannot treat the A like a B, because this exactly breaks down in the list case. Thus, if I have a background task (or other thingy) acting on a B, I cannot feed an A into it.

I think I fully understand your point now. My arguments were based too much on the list case and not on this generic approach.

> What is your understanding of covariant and contravariant? As far as I understand, List is neither covariant nor contravariant, hence the problem. pop is covariant. If I have a list that contains As and pop values from it, I could treat the As as Bs if A.isExtensionOf(B). push on the other hand is contravariant. If I have a List of Bs I could push As into it if A.isExtensionOf(B), which is exactly the other way round. If I had a list of As I cannot push Bs into it, even if A.isExtensionOf(B), because an A is somehow more than a B.

Yes, I agree with your definition of co- and contravariant as well as your point here.

To sum up: A (List) that depends on some generic T (Animal) can never implement isSubclassOf with respect to the generic T because methods of A might act co- or contravariant concerning T.

See: 

List<'Animal>.isSubtypeOf(List<'Dog>) == True => List<'Animal>.pop() should give us a Dog. => Contradiction!

List<'Dog>.isSubtypeOf(List<'Animal>) == True => List<'Dog>.push(...) should accept an Animal. => Contradiction!

This means that: A<'T>.isSubtypeOf(A<'U>) == False; For all T !=U. (This was a proof by example but should hold for the general case).

Okay then what about A<? extends T> case? Or in our case List<? extends Animal>. This generalizes the generic to anything that extends Animal. So let's try it:

List<'Dog>.isSubtypeOf(List<? extends Animal>) == True. List<? extends Animal>.push(...) accepts a Cat as it also extends animal. But List<'Dog> does not accept a Cat thus isSubtype == False => Contradiction.

Thus the <? extends..> does not solve this Problem. The only Problem that this solves is that it marks the input as "dangerous". So if you have a method that accepts x of Type A<? extends T> (implemented as mentioned above) you CANNOT invoke any contravariant methods on x.

We have no notion on whether a method acts co- or contravariant on its input or output and I don't want to introduce any, thus we should only allow A<'T>.isSubtypeOf(A<'U>) == true; iff T==U. But then a job that can be invoked on a list of ids cannot be invoked on a list of user ids and this would probably be a nice use-case.

Do you have any solution?

> And also nice to have a discussion on this level here.
It certainly is! Even though it feels like there should be a go-to solution. Wikipedia tells that most implementations use just covariant-y behaviour for our isSubtypeOf/isExtensionOf even though this might be unsound but easier to reason about for programmers. No idea what "most" means here.

Best
Oskar

klees replied:
Hi Oskar!

I'm happy we sorted that one out =) Indeed, List<? extends B> is not sound as well.

Possible solutions to the problem seem to be:
* Split covariant and contravariant part of the list. Go on as you proposed with these two separated. I think for your usecase the `pop`-part could be sufficient. Could be hard to find names, though, and won't help if arrays are in fact passed.
* Drop this whole notion of `extension` or `subclass`. AFAIK this is the source of the trouble. You want to able to stuff a list of courses in a place where a list of objects is processed? Add an explicit transformation step.
* Keep this whole notion of `extension` or `subclass`. Don't lie about it for the list. You want to able to stuff a list of courses in a place where a list of objects is processed? Add an explicit transformation step.

I think I would go with the second one. I somehow doubt, that all inheritance-hierarchies in ILIAS respect LSP, so the last approach will be doomed anyway. In general, "implementing some interface" to me seems to be a saner concept than "inheriting from some other class". The former seems to say more about expected behaviour, while the latter seems to be mostly used as a way to share code (in general, not only in ILIAS). There might be some solution for your course/object-case hidden in the interface approach.

Best regards!

otruffer commented:
@Amstutz: I see the benefits for a factory, if you just want an implementation of an interface, depending on the input values. e.g. $repositoryFactory->getObject($object_id) will give you an ilObjCourse or an ilObjUser depending on what specific object is saved behind the object id (abstract factory). Or if you have a moderately complicated construction process. Here neither of these are the case. What exactly is the benefit of a factory here? 

The only one I see is, if the classes have some dependencies, they can be injected via the factory. Do you think this is necessary here?

otruffer commented:
@klees Okay I think i've got it.
1. As you said it would be strange to lie about the subclassing, so a List<'A> is only a subclass of List<'B> iff 'A equals 'B. 

2. For my use case I want to be able to pass **'B** : List<'B> to a method that accepts List<'A> where 'B extends' A. I can implicitly (instead of explicitly as you suggested) convert the **'B** : List<'B> to a **'A** : List<'A> by just putting all elements **b** of **'B** into **'A**. So the only method i need is **push** which is covariant for every list and thus the approach hold for arbitrary types where 'B extends 'A. 

Thus we have both, soundness and simplicity of use. Is there anything that i missed?

klees reviewed (commented):
Hey @otruffer, sounds good to me. Thx!

Amstutz commented:
Can this be merged?
